src/main/scala/ie/zalando/fabric/gateway/service/ZeroDowntimeIngressTransitions.scala

```diff
@@ -0,0 +1,24 @@
+package ie.zalando.fabric.gateway.service
+
+import ie.zalando.fabric.gateway.models.SynchDomain.{GatewayMeta, GatewaySpec, IngressDefinition}
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class ZeroDowntimeIngressTransitions(ingressDerivationChain: IngressDerivationChain) {
+
+  def defineSafeRouteTransition(spec: GatewaySpec, metadata: GatewayMeta, existingRoutes: Seq[IngressDefinition])(
+      implicit ec: ExecutionContext): Future[List[IngressDefinition]] = {
+    ingressDerivationChain
+      .deriveRoutesFor(spec, metadata)
+      .map { desiredRoutes =>
+        val deleted = relativeComplement(desiredRoutes, existingRoutes)
+        val created = relativeComplement(existingRoutes, desiredRoutes)
+        if (created.isEmpty) desiredRoutes else desiredRoutes ++ deleted
+      }
+  }
+
+  private def relativeComplement(a: Seq[IngressDefinition], b: Seq[IngressDefinition]) =
```

Comment:
maybe more descriptive names than a and b?

Reply:
It's basically the exact set theory definition. 
<img width="678" alt="Screenshot 2020-07-21 at 16 14 19" src="https://user-images.githubusercontent.com/868640/88072694-42c06400-cb6d-11ea-9ef6-0f91bde922fa.png">

Reply:
eeep... so we're naming things mathematically now?

Reply:
For pure set theory operations, absolutely I am. Scala already has `union`, `intersection` and `diff` defined, I can't imagine how ugly things would be if everyone invented their own terminology for Set Theory operations.